### PR TITLE
ISPN-12667 GlobalState incompatibility between 11.x and 12.x

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-12667

Backport: https://github.com/infinispan/infinispan/pull/9020

### CODE REVIEW

File: core/src/main/java/org/infinispan/distribution/ch/impl/ReplicatedConsistentHash.java [hunk 1]
```diff
@@ -125,7 +125,8 @@ public ReplicatedConsistentHash union(ReplicatedConsistentHash ch2) {
 
    private static List<Address> parseMembers(ScopedPersistentState state, String numMembersPropertyName,
                                              String memberPropertyFormat) {
-      int numMembers = Integer.parseInt(state.getProperty(numMembersPropertyName));
+      String property = state.getProperty(numMembersPropertyName);
+      int numMembers = property == null ? 0 : Integer.parseInt(property);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/distribution/ch/impl/ReplicatedConsistentHash.java:129]: nitpick :) 
```suggestion
      String property = state.getProperty(numMembersPropertyName);
      if (property == null) {
          return Collections.emptyList();
      }
      int numMembers = Integer.parseInt(property);
```

File: core/src/test/java/org/infinispan/globalstate/GlobalStateBackwardsCompatibilityTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+/**
+ * ISPN-12667 A test to ensure that global state properties from previous Infinispan majors are compatible.
+ *
+ * @author Ryan Emerson
+ * @since 12.0
+ */
+public class GlobalStateBackwardsCompatibilityTest extends MultipleCacheManagersTest {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/globalstate/GlobalStateBackwardsCompatibilityTest.java:27]: The test isn't running.
```suggestion
@Test(groups = "functional", testName = "globalstate.GlobalStateBackwardsCompatibilityTest")
public class GlobalStateBackwardsCompatibilityTest extends MultipleCacheManagersTest {
```

File: core/src/test/java/org/infinispan/globalstate/GlobalStateBackwardsCompatibilityTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      createCacheManagerWithGlobalState(MEMBER_0, tmpDirectory(this.getClass().getSimpleName(), "0"));
+      createCacheManagerWithGlobalState(MEMBER_1, tmpDirectory(this.getClass().getSimpleName(), "1"));
+      waitForClusterToForm(CACHE_NAME);
+   }
+
+   @AfterClass(alwaysRun = true)
+   protected void clearTempDir() {
+      Util.recursiveFileRemove(tmpDirectory(this.getClass().getSimpleName()));
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/globalstate/GlobalStateBackwardsCompatibilityTest.java:41]: -1 to delete the folder in a separate method.
It is being invoked before stopping the cache manager and I'm getting `Caused by: java.io.FileNotFoundException: /home/pedro/Work/infinispan/core/target/infinispanTempFiles/GlobalStateBackwardsCompatibilityTest/1/___global.state (No such file or directory)`

IMO, it is better to overwrite `destroy()`, stopping the cache manager first, and then remove the folder.
```
   @AfterClass(alwaysRun = true)
   @Override
   protected void destroy() {
      super.destroy();
      Util.recursiveFileRemove(tmpDirectory(this.getClass().getSimpleName()));
   }
```

**Review by pruivo — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/infinispan/globalstate/GlobalStateBackwardsCompatibilityTest.java (hunk 2 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/globalstate/GlobalStateBackwardsCompatibilityTest.java:28]: I swear I forgot the `@Test` annotation every time! Or copy and paste and forget to update the `testName` :face_with_head_bandage:

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
integrated! thanks @ryanemerson !

**Pull request merged**